### PR TITLE
Replace assertions in the code

### LINKED ISSUE #33: Replace assertions by informative errors
Every non-recoverable errors due to wrong inputs in the `tqec` Python package are asserted using `assert`. Even though the assertion message is better than nothing, it would be clearer to raise errors with descriptive names and descriptions, that can potentially be `catch`ed by a power-user expecting such an error and knowing how to recover from it.

### PR DESCRIPTION
**Pull request by nelimee**
Fixes #33

### CODE REVIEW

File: src/tqec/detectors/gate.py [hunk 1]
```diff
@@ -3,8 +3,33 @@
 
 import cirq
 import stim
-
 from tqec.detectors.measurement_map import CircuitMeasurementMap
+from tqec.errors import QubitTypeError
+from tqec.errors import MeasurementAppliedOnMultipleQubits
+
+
+class GlobalMeasurementLoopbackMissing(Exception):
```
[the remark this answers is not in the document]
Comment (nelimee) [reply src/tqec/detectors/gate.py:11]: You are right, this is a typo, also I am not native either so I may (and I do often) make mistakes.

**Comment by inmzhang** [comment]
Refer to the structure defined in [qiskit](https://github.com/search?q=repo%3AQiskit%2Fqiskit%20exceptions.py&type=code), there is a [`QiskitError`](https://github.com/Qiskit/qiskit/blob/3364ba50230805b99ce4166f3430856ef5b59f36/qiskit/circuit/exceptions.py) used as the base exception for the whole project. And every "domain"(sub-package) has its own exceptions defined as the children, e.g. [`CircuitError`](https://github.com/Qiskit/qiskit/blob/3364ba50230805b99ce4166f3430856ef5b59f36/qiskit/circuit/exceptions.py).

For the most use cases, using the base exception class to catch error should be enough. So I think distributing the exception definitions in a structured manner should be fine. The problem is that whether we need a separate `exceptions.py` file for each "domain" or just define them directly within the modules.

**Comment by nelimee** [comment]
Right, good observation. Let's follow the same principle and have a base `TQECException` class defined in `tqec.errors` (soon renamed to `tqec.exceptions`), from which all the exception will inherit.

Also, I prefer to have exceptions defined right where they are raised, but I do not have any argument in either way, so both are fine to me.

File: src/tqec/detectors/gate.py [hunk 2]
```diff
@@ -4,18 +4,22 @@
...
-from tqec.exceptions import MeasurementAppliedOnMultipleQubitsException, QubitTypeException
+from tqec.exceptions import (
+    MeasurementAppliedOnMultipleQubitsException,
+    QubitTypeException,
+    TQECException,
+)
 
 
-class GlobalMeasurementLookbackMissingException(Exception):
+class GlobalMeasurementLookbackMissingException(TQECException):
```
Comment (inmzhang) [inline src/tqec/detectors/gate.py:14]: Some exceptions seem to be a simple wrapper around `TQECException` and I wonder whether we need to define those too-detailed exceptions. For example, `GlobalMeasurementLookbackMissingException()` can be replaced with the `TQECException("Global measurement lookback offsets have not been computed...")`. And this is also what `qiskit` did for some of its exceptions with `QiskitError("...")`.

Or perhaps we need the perspectives of others on this issue.

**Comment by nelimee** [comment]
Seems like I forgot to merge that one. Just to be sure, @inmzhang do you agree with the current state, or should I modify a few things?

**Comment by inmzhang** [comment]
Currently, what is the criterion of determining when to create a specific exception type? I saw the exceptions like `DimensionNotEvenException`, which seems can also be replaced with the general `TQECException`.

**Comment by nelimee** [comment]
> Currently, what is the criterion of determining when to create a specific exception type?

If the exception can be raised from several places (e.g., `GlobalMeasurementLookbackMissingException` or `DimensionNotEvenException`) or if formatting the exception message requires non-trivial or long code (e.g. `CannotComputeMeasurementOffsetLookbackException` or the sub-classes of `ScheduleException`) I used sub-classes inheriting from `TQECException`.

For the specific case of `DimensionNotEvenException`, I prefer to have a standalone `Exception` sub-class defined than to copy-paste the exact same error message on two places. 

With the explanations above, do you think I should change some of the exceptions?

**Comment by inmzhang** [comment]
That makes sense to me and I think it's ok to keep this unchanged.

**Pull request merged**